### PR TITLE
ARROW-16158: [C++][R] Rename ARROW_ENGINE to ARROW_SUBSTRAIT

### PR DESCRIPTION
**Pull request by westonpace**
At the moment there is not enough non-Substrait functionality to justify a dedicated engine module.  In the future there likely will be and so I have some of the general structure around in the C++ (e.g. `ARROW_ENGINE_EXPORT`) but removed any external facing references.

### CODE REVIEW

**Review by nealrichardson — COMMENTED** [review]
R changes make sense to me, just one comment on the cmake

File: r/NAMESPACE
Comment (kou) [inline r/NAMESPACE:208]: Could you keep this list in alphabetical order?

**Comment by kou** [comment]
@github-actions crossbow submit -g nightly

File: cpp/CMakeLists.txt
Comment (westonpace) [reply cpp/CMakeLists.txt:356]: Yes.  In the future we could presumably add conditional-feature flags.  For example, if a Substrait file says the format is Parquet and the consumer was built without Parquet then reject the plan as invalid.  I created ARROW-16274 to track this.

**Review by kou — APPROVED** [review]
+1

(I think that https://github.com/apache/arrow/pull/12915/files/ec22d466b469202d27df76f5445809b77e978174#diff-bae2e9b731f293cf61d77c971d4152d31b9d49b0354817931e5a51c6c27ecf19 should be fixed before we merge this.)

**Comment by nealrichardson** [comment]
I fixed the R namespace and rebased to resolve conflicts, this should be good to merge once CI finishes

**Comment by kou** [comment]
@westonpace I don't know why but C++ on s390x test on Travis CI is failed since this commit: https://app.travis-ci.com/github/apache/arrow/builds

Could you confirm it?

https://app.travis-ci.com/github/apache/arrow/jobs/568028126

```text
The following tests FAILED:
	 49 - arrow-dataset-file-parquet-test (Failed)
	 59 - parquet-internals-test (Failed)
	 60 - parquet-reader-test (Failed)
	 62 - parquet-arrow-test (Failed)
	 63 - parquet-arrow-internals-test (Failed)
	 64 - parquet-encryption-test (Failed)
```

**Comment by kou** [comment]
ARROW-16305

**Pull request closed without merging**
